internal/impl/snowflake/output_snowflake_streaming.go

```diff
@@ -456,6 +541,129 @@ func (o *snowflakeStreamerOutput) WriteBatch(ctx context.Context, batch service.
 	return err
 }
 
+type schemaMigrationNeededError struct {
+	migrator func(ctx context.Context) error
+}
+
+func (schemaMigrationNeededError) Error() string {
+	return "schema migration was required and the operation needs to be retried after the migration"
+}
+
+func (o *snowflakeStreamerOutput) MigrateMissingColumn(ctx context.Context, col streaming.MissingColumnError) error {
+	o.schemaMigrationMu.Lock()
+	defer o.schemaMigrationMu.Unlock()
+	msg := service.NewMessage(nil)
+	msg.SetStructuredMut(map[string]any{
+		"name":  col.RawName(),
+		"value": col.Value(),
+	})
+	out, err := msg.BloblangQuery(o.schemaEvolutionMapping)
+	if err != nil {
+		return fmt.Errorf("unable to compute new column type for %s: %w", col.ColumnName(), err)
+	}
+	v, err := out.AsBytes()
+	if err != nil {
+		return fmt.Errorf("unable to compute new column type for %s: %w", col.ColumnName(), err)
+	}
+	columnType := string(v)
+	if err := validateColumnType(columnType); err != nil {
+		return err
+	}
+	o.logger.Infof("identified new schema - attempting to alter table to add column: %s %s", col.ColumnName(), columnType)
+	_, err = o.restClient.RunSQL(ctx, streaming.RunSQLRequest{
+		// This looks very scary and it *should*. This is prone to SQL injection attacks. The column name is
+		// quoted according to the rules in Snowflake's documentation. This is also why we need to
+		// validate the data type, so that you can't sneak an injection attack in there.
+		Statement: fmt.Sprintf(`ALTER TABLE IDENTIFIER(?)
+    ADD COLUMN IF NOT EXISTS %s %s
+      COMMENT 'column created by schema evolution from Redpanda Connect'`,
+			col.ColumnName(),
+			columnType,
+		),
+		// Currently we set of timeout of 30 seconds so that we don't have to handle async operations
+		// that need polling to wait until they finish (results are made async when execution is longer
+		// than 45 seconds).
+		Timeout:  30,
+		Database: o.db,
+		Schema:   o.schema,
+		Role:     o.role,
+		Bindings: map[string]streaming.BindingValue{
+			"1": {Type: "TEXT", Value: o.table},
+		},
+	})
```

Comment:
Could we have a wrapper func so we don't duplicate this code & comments in `MigrateNotNullColumn()`?

Reply:
Yes - thanks it's much easier to read now :)